File: learn/sorting/quick-sort/quick_sort.go


type QuickSort struct{}

func (s *QuickSort) Sort(arr []int) {
	if len(arr) > 0 {
		s.quickSort(arr, 0, len(arr)-1)
	}
}

func (s *QuickSort) quickSort(arr []int, low, high int) {
	if low < high {
		pi := s.partition(arr, low, high)
		s.quickSort(arr, low, pi-1)
		s.quickSort(arr, pi+1, high)
	}
}

func (s *QuickSort) partition(arr []int, low, high int) int {
	pivot := arr[high]
	i := low - 1

	for j := low; j < high; j++ {
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	arr[i+1], arr[high] = arr[high], arr[i+1]
	return i + 1
}

File: learn/sorting/quick-sort/quick_sort.rs
pub struct QuickSort;

impl QuickSort {
    pub fn sort(arr: &mut [i32]) {
        let n = arr.len();
        if n > 0 {
            Self::quick_sort(arr, 0, (n - 1) as isize);
        }
    }

    fn quick_sort(arr: &mut [i32], low: isize, high: isize) {
        if low < high {
            let pi = Self::partition(arr, low, high);
            Self::quick_sort(arr, low, pi - 1);
            Self::quick_sort(arr, pi + 1, high);
        }
    }

    fn partition(arr: &mut [i32], low: isize, high: isize) -> isize {
        let pivot = arr[high as usize];
        let mut i = low - 1;

        for j in low..high {
            if arr[j as usize] < pivot {
                i += 1;
                arr.swap(i as usize, j as usize);
            }
        }
        arr.swap((i + 1) as usize, high as usize);
        i + 1
    }
}
